ipywebrtc/webrtc.py

```diff
@@ -278,74 +339,132 @@ def save(self, filename=None):
         with open(filename, 'wb') as f:
             f.write(self.data)
 
-    _video_src = Unicode('').tag(sync=True)
-
-
-
-# monkey patch, same as https://github.com/jupyter-widgets/ipywidgets/pull/2146
-if 'from_json' not in widgets.Image.value.metadata:
-    widgets.Image.value.metadata['from_json'] = lambda js, obj: None if js is None else js.tobytes()
 
 @register
-class MediaImageRecorder(DOMWidget):
-    """Creates a recorder which allows to grab an Image from a MediaStream widget.
+class VideoRecorder(Recorder):
+    """Creates a recorder which allows to record a MediaStream widget, play the
+    record in the Notebook, and download it or turn it into a Video widget.
     """
     _model_module = Unicode('jupyter-webrtc').tag(sync=True)
     _view_module = Unicode('jupyter-webrtc').tag(sync=True)
-    _model_name = Unicode('MediaImageRecorderModel').tag(sync=True)
-    _view_name = Unicode('MediaImageRecorderView').tag(sync=True)
+    _model_name = Unicode('VideoRecorderModel').tag(sync=True)
+    _view_name = Unicode('VideoRecorderView').tag(sync=True)
     _view_module_version = Unicode(semver_range_frontend).tag(sync=True)
     _model_module_version = Unicode(semver_range_frontend).tag(sync=True)
 
-    stream = Instance(MediaStream, allow_none=True,
-         help=MediaRecorder.stream.metadata['help'])\
-        .tag(sync=True, **widget_serialization)
-    image = Instance(Image, help='An instance of ipywidgets.Image that will receive the grabbed image.',
-        allow_none=True).tag(sync=True, **widget_serialization)
-    filename = Unicode('recording', help=MediaRecorder.filename.metadata['help']).tag(sync=True)
-    autosave = Bool(False, help=MediaRecorder.autosave.metadata['help'])
+    @validate('stream')
+    def _valid_stream(self, proposal):
+        stream = proposal['value']
+        if not (isinstance(stream, VideoStream) or
+                isinstance(stream, WidgetStream) or
+                isinstance(stream, CameraStream) or
+                isinstance(stream, ImageStream)):
+            raise TraitError('Cannot record a video from {} instance'.format(
```

Comment:
Why whitelist, why not `isinstance(stream, MediaStream)`?

Reply:
I think the check should be done on the frontend, there we know if a MediaStream contains an audio or videotrack.

Reply:
I've put this because it doesn't make sense to create a VideoRecorder for an AudioStream

Reply:
But this will not allow extending by 3rd parties, you could say create classes that 'mark' the mediastream as having video and/or audio, or have static properties. I'd say, just leave the checking at the frontend.

Reply:
agree